torch/_functorch/aot_autograd.py

```diff
@@ -908,34 +958,89 @@ def joint_forward_backward(
+        assert len(meta.input_info) == len(primals)
+        outs = forward_or_joint(primals_before_cloning, primals, maybe_tangents)
+        return outs
+
+    def flat_fn_no_input_mutations(primals: List[Any], maybe_tangents: Optional[List[Any]]):
+        # When tracing the joint fwd + bwd, making sure to clone any inputs that are mutated first.
+        # We need to ensure that the inputs we pass to autograd.grad() are the *original*
+        # inputs, and not their mutated values.
+        if maybe_tangents is not None:
+            primals_after_cloning = [
+                maybe_to_fresh_input(i, t) for i, t in enumerate(primals)
+            ]
         else:
-            return t
-
-    def from_fun(t):
-        if not isinstance(t, Tensor) or not torch._is_functional_tensor(t):
-            return t
-        torch._sync(t)
-        return torch._from_functional_tensor(t)
-
-    def functionalized_joint(
-        primals: List[Any], tangents: List[Any]
-    ) -> Tuple[List[Any], List[Any]]:
+            primals_after_cloning = primals
+        outs = flat_fn_with_synthetic_bases_expanded(primals, primals_after_cloning, maybe_tangents)
+        return outs
 
+    def functionalized_f_helper(primals_and_maybe_tangents):
+        if trace_joint:
+            primals, maybe_tangents = primals_and_maybe_tangents
+        else:
+            primals = primals_and_maybe_tangents
+            maybe_tangents = None
         # Wrap inputs into functional wrappers
-        f_primals, f_tangents = pytree.tree_map(to_fun, (primals, tangents))
+        f_primals = pytree.tree_map(to_fun, primals)
+        f_tangents = None if maybe_tangents is None else pytree.tree_map(to_fun, maybe_tangents)
         torch._enable_functionalization(reapply_views=True)
         try:
             # Run the joint
-            f_outs = joint_forward_backward(f_primals, f_tangents)
+            f_outs = flat_fn_no_input_mutations(f_primals, f_tangents)
         finally:
             torch._disable_functionalization()
 
+        if keep_input_mutations:
+            # Note: This is a bit annoying. There's a layering issue here, where:
+            # (1) functionalization needs to operate on **synthetic base** inputs, before unpacking them into the "real" inputs.
+            # (2) For keep_input_mutations, we support tracing a call to copy_() directly on mutated inputs.
+            #     However, we **only** want to support this for inputs that have data-only (and no metadata) mutations,
+            #     because inductor (and backends in generally) would prefer not to see these (e.g. as_strided_(), resize_()).
+            #     This makes it pretty difficult for this logic to operate on synthetic bases.
+            #     For example, suppose "x" and "y" are aliased inputs, and we witness:
+            #       x.add_(1)
+            #       y.t_()
+            #     We don't actually want the compiler to witness a copy_() on the synthetic base here: it should
```

Comment:
The copy on synthetic base is akin to directly updating the storage. Maybe it is marginally better to tell the compiler about mutation on x, but updating the storage also seems sound to me?

Reply:
I guess this is what you say in the comment below. Your example on GH is much better than the example in the code here

Reply:
Yes agreed, I'll put that directly in the code comment